File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        curr = dummy
        N = 0 

        while curr.next:
            N += 1
            curr = curr.next


        cnt = 0
        curr = dummy
        while curr.next:
            if cnt == (N - n):
                curr.next = curr.next.next
                break
            else:
                curr = curr.next
                cnt += 1

        return dummy.next            

        

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) return null;
        return mergeHelper(lists, 0, lists.length - 1);
    }

    private ListNode mergeHelper(ListNode[] lists, int l, int r) {
        if (l == r) return lists[l];
        if (l > r) return null;

        int m = (l + r) / 2;
        ListNode n1 = mergeHelper(lists, l, m);
        ListNode n2 = mergeHelper(lists, m + 1, r);

        return mergeTwoLists(n1, n2);
    }

    public ListNode mergeTwoLists(ListNode node1, ListNode node2) {
        ListNode dummy = new ListNode(-1);
        ListNode current = dummy;

        while (node1 != null && node2 != null) {
            if (node1.val <= node2.val) {
                current.next = node1;
                node1 = node1.next;
            } else {
                current.next = node2;
                node2 = node2.next;
            }
            current = current.next;
        }
        current.next = (node1 != null) ? node1 : node2;
        return dummy.next;
    }
}
